### PR TITLE
Bidirectional and dynamic mounts on VM

### PR DESCRIPTION
**Pull request by Hoxo**

### CODE REVIEW

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java [hunk 1]
```diff
@@ -228,6 +234,49 @@ private StepResult allocateVm() {
...
+        if (mountPodAllocated) {
+            return StepResult.ALREADY_DONE;
+        }
+
+        log().info("{} Allocating mount pod for VM {}", logPrefix(), vm.vmId());
+        try {
+            mountHolder = allocationContext.mountHolderManager().allocateMountHolder(vm.spec());
+            mountPodAllocated = true;
+        } catch (KubernetesClientException e) {
+            log().error("{} Cannot allocate mount holder for vm {}: {}", logPrefix(), vm.vmId(), e.getMessage());
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java:251]: already exists?

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java [hunk 2]
```diff
@@ -228,6 +234,49 @@ private StepResult allocateVm() {
...
+
+        if (mountPodSet) {
+            return StepResult.ALREADY_DONE;
+        }
+
+        var pod = mountHolder.podName();
+        log().info("{} Setting mount pod name {} for VM", logPrefix(), pod);
+        try {
+            withRetries(log(), () -> allocationContext.vmDao().setMountPod(vm.vmId(), pod, null));
+            mountPodSet = true;
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java:270]: redundant variable
dublicated by `vm.withMounfPod(pod)` (line 276)

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocationContext.java [hunk 3]
```diff
@@ -83,4 +93,40 @@ public DeleteVmAction createDeleteVmAction(Vm vm, String description, String req
 
         return new DeleteVmAction(vm, deleteOp.id(), this);
     }
+
+    public Pair<UnmountDynamicDiskAction, Operation> createUnmountAction(@Nullable Vm vm, DynamicMount dynamicMount,
+                                                             @Nullable TransactionHandle tx) throws SQLException
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocationContext.java:98]: broken indent

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package ai.lzy.allocator.alloc.dao;
+
+import ai.lzy.allocator.model.DynamicMount;
+import ai.lzy.model.db.TransactionHandle;
+import org.jetbrains.annotations.Nullable;
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:5]: we use `jakarta` annonations

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Nullable
+    DynamicMount get(String id, boolean forUpdate, @Nullable TransactionHandle tx) throws SQLException;
+
+    void delete(String id, @Nullable TransactionHandle tx) throws SQLException;
+
+    @Nullable
+    DynamicMount update(String id, DynamicMount.Update update, @Nullable TransactionHandle tx) throws SQLException;
+
+    long countForVolumeClaimName(String clusterId, String volumeClaimName, @Nullable TransactionHandle tx)
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:21]: count usages?

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/AllocatorService.java [hunk 6]
```diff
@@ -83,11 +102,20 @@ public AllocatorService(VmDao vmDao, @Named("AllocatorOperationDao") OperationDa
         this.allocationContext = allocationContext;
         this.config = config;
         this.cacheConfig = cacheConfig;
+        this.mountConfig = mountConfig;
 
         restoreRunningActions();
     }
 
     private void restoreRunningActions() {
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/services/AllocatorService.java:110]: you may wish to introduce a new component "RecoverOperations" and move this logic there

File: lzy/allocator/src/main/java/ai/lzy/allocator/util/DaoUtils.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package ai.lzy.allocator.util;
+
+public final class DaoUtils {
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/util/DaoUtils.java:3]: you may with to move it to the `utils-db` module

File: lzy/allocator/src/main/java/ai/lzy/allocator/util/KuberUtils.java [hunk 8]
```diff
@@ -11,4 +11,12 @@ private KuberUtils() {
     public static boolean isResourceAlreadyExist(Exception e) {
         return e instanceof KubernetesClientException ex && ex.getCode() == HttpURLConnection.HTTP_CONFLICT;
     }
+
+    public static boolean isNotRetryable(KubernetesClientException e) {
+        var code = e.getCode();
+        if (code == 429) { // Too many requests
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/util/KuberUtils.java:17]: `io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.HTTP_TOO_MANY_REQUESTS`  :)

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorApiTestBase.java [hunk 9]
```diff
@@ -291,7 +291,7 @@ protected void registerVm(String vmId, String clusterId) {
                 });
             } catch (StatusRuntimeException e) {
                 if (e.getStatus().getCode() == Status.Code.FAILED_PRECONDITION) {
-                    System.err.println("Fail to register VM %s: %s".formatted(vmId, e.getStatus().toString()));
+                    System.err.printf("Fail to register VM %s: %s%n", vmId, e.getStatus());
```
Comment (imakunin) [inline lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorApiTestBase.java:294]: let's replace with logger

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    @Override
+    protected void updateStartupProperties(Map<String, Object> props) {
+        super.updateStartupProperties(props);
+        props.put("allocator.allocation-timeout", "30s");
+        props.put("allocator.kuber-tunnel-allocator.enabled", "true");
+        props.put("allocator.mount.enabled", "true");
+        props.put("allocator.mount.podImage", "ubuntu");
+        props.put("allocator.mount.workerMountPoint", "/mnt/worker");
```
Comment (imakunin) [inline lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java:49]: let's move all temp data to the `'/tmp` dir

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        props.put("allocator.mount.hostMountPoint", "/mnt/host");
+    }
+
+    @Test
+    public void allocatorServiceShouldValidateEmptyMountRequest() {
```
Comment (imakunin) [inline lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java:54]: too long name, just `emptyMountRequst`

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        var exception = Assert.assertThrows(StatusRuntimeException.class,
+            () -> authorizedAllocatorBlockingStub.mount(VmAllocatorApi.MountRequest.getDefaultInstance()));
+        Assert.assertEquals(Status.INVALID_ARGUMENT.getCode(), exception.getStatus().getCode());
+        Assert.assertTrue(exception.getMessage().contains("vm_id isn't set"));
+        Assert.assertTrue(exception.getMessage().contains("mount_path isn't set"));
+        Assert.assertTrue(exception.getMessage().contains("volume_type isn't set"));
+    }
+
+    @Test
+    public void allocatorServiceShouldValidateVm() {
```
Comment (imakunin) [inline lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java:64]: just `invalidVm()`

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java (hunk 4 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:5]: Why? Jetbrains annotations are tracked by IDEA and it helps to highlight possible NPE. But I don't see much benefit from jakarta/javax annotations

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java (hunk 5 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:21]: For me, it sounds too ambiguous. _Count usages_ for what?

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java (hunk 10 above)
Comment (Hoxo) [reply lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java:49]: These are just stubs with no meaning. These values won't be used to actually mount something in test env

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java (hunk 4 above)
Comment (imakunin) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:5]: jakarta annotations do the same (just configure IDE) :)

Comment (imakunin) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:5]: it seems that we will have to abandon the use of JetBrains IDE

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java (hunk 5 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java:21]: Yeah, the main idea is to count how many mounts use specific volume claim. Maybe it should be named more clearly

File: lzy/allocator/src/main/java/ai/lzy/allocator/util/KuberUtils.java (hunk 8 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/util/KuberUtils.java:17]: The word `internal` implies that we should use it. I would prefer using own constants

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java (hunk 1 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java:251]: `allocateMountHolder` already resolves it

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java (hunk 12 above)
Comment (Hoxo) [reply lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java]: done

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java (hunk 11 above)
Comment (Hoxo) [reply lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorServiceMountsTest.java]: done

File: lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorApiTestBase.java (hunk 9 above)
Comment (Hoxo) [reply lzy/allocator/src/test/java/ai/lzy/allocator/test/AllocatorApiTestBase.java]: done

File: lzy/allocator/src/main/java/ai/lzy/allocator/util/KuberUtils.java (hunk 8 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/util/KuberUtils.java]: Replaced it with constant

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java (hunk 4 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/dao/DynamicMountDao.java]: Returned jakarta annotations

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java (hunk 2 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/alloc/AllocateVmAction.java]: Fixed

File: lzy/allocator/src/main/java/ai/lzy/allocator/util/DaoUtils.java (hunk 7 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/util/DaoUtils.java]: Moved it

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/AllocatorService.java (hunk 6 above)
Comment (Hoxo) [reply lzy/allocator/src/main/java/ai/lzy/allocator/services/AllocatorService.java]: Moved to `RestoreOperations`

**Review by imakunin — APPROVED** [review]

**Comment by imakunin** [comment]
LGTM =)

**Pull request merged**